clover_runtime.rb

```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+require "jwt"
+
+class CloverRuntime < Roda
+  include CloverBase
+
+  plugin :default_headers, "Content-Type" => "application/json"
+
+  plugin :hash_branches
+  plugin :json
+  plugin :all_verbs
+  plugin :json_parser
+
+  autoload_routes("runtime")
+
+  plugin :error_handler do |e|
+    error = parse_error(e)
+
+    {error: error}.to_json
+  end
+
+  def get_jwt_payload(request)
+    return unless (v = request.env["HTTP_AUTHORIZATION"])
+    jwt_token = v.sub(%r{\ABearer:?\s+}, "")
+    begin
+      JWT.decode(jwt_token, Config.clover_runtime_token_secret, true, {algorithm: "HS256"})[0]
+    rescue JWT::DecodeError
+    end
+  end
+
+  route do |r|
+    if (jwt_payload = get_jwt_payload(r)).nil? || (@vm = Vm.from_ubid(jwt_payload["sub"])).nil?
+      fail CloverError.new(401, "InvalidCredentials", "There was an error logging in")
```

Comment:
should we return `invalid JWT format or claim in Authorization header` instead of login error?

Reply:
see, spec/routes/api/auth_spec.rb

```rb
  it "wrong jwt" do
    header "Authorization", "Bearer wrongjwt"
    get "/api/project"

    expect(last_response.status).to eq(400)
    expect(JSON.parse(last_response.body)["error"]["message"]).to eq("invalid JWT format or claim in Authorization header")
    expect(JSON.parse(last_response.body)["error"]["type"]).to eq("InvalidRequest")
  end
```

Reply:
Should we return distinct messages for an invalid JWT and a non-existent VM, or should we use the same message? I tend not to be overly explicit in error messages to prevent data exposure. I could change the error message to  `invalid JWT format or claim in Authorization header` for the both cases.

Interestingly, our APIs return an HTTP 400 for an invalid JWT instead of a 401. I expected a 401 response for all JWT-related failures.